### ISSUE TITLE
"Fix all" does nothing sometimes and crashes

### ISSUE THREAD

**Issue by davkean** [dotnet/roslyn#16086]
This is a branch from https://github.com/dotnet/roslyn/issues/15762 - this is a case where Fix All just doesn't run (this is after I have the fix for https://github.com/dotnet/roslyn/issues/15761) . I don't have a 100% repro on this, but it occurs after running a few fixes, here's a video to show I'm not crazy:

https://vid.me/VPXO

Note, this time it crashed, which I've never seen before - the stack was:

```
Application: devenv.exe
Framework Version: v4.0.30319
Description: The application requested process termination through System.Environment.FailFast(string message).
Message: System.InvalidOperationException: Sequence contains no elements
   at System.Linq.Enumerable.First[TSource](IEnumerable`1 source)
   at System.Linq.ImmutableArrayExtensions.First[T](ImmutableArray`1 immutableArray)
   at Microsoft.CodeAnalysis.CodeFixes.BatchSimplificationFixAllProvider.<AddDocumentFixesAsync>d__2.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.CodeAnalysis.CodeFixes.BatchFixAllProvider.<GetFixAsync>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.CodeAnalysis.CodeFixes.BatchFixAllProvider.<GetFixAsync>d__2.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.FixAllGetFixesService.<GetFixAllCodeActionAsync>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.FixAllGetFixesService.<GetFixAllOperationsAsync>d__2.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.CodeAnalysis.CodeActions.CodeAction.<GetOperationsCoreAsync>d__16.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.CodeAnalysis.CodeActions.CodeAction.<GetOperationsAsync>d__15.MoveNext()
Stack:
   at System.Environment.FailFast(System.String, System.Exception)
   at Microsoft.CodeAnalysis.FailFast.OnFatalException(System.Exception)
   at Microsoft.CodeAnalysis.ErrorReporting.FatalError.Report(System.Exception, System.Action`1<System.Exception>)
   at Microsoft.VisualStudio.IntegrationTest.Setup.TestExtensionManager+ExtensionManager.HandleException(System.Object, System.Exception)
   at Microsoft.VisualStudio.IntegrationTest.Setup.TestExtensionManager+ExtensionManager.CanHandleException(System.Object, System.Exception)
   at Microsoft.CodeAnalysis.Extensions.IExtensionManagerExtensions.PerformAction(Microsoft.CodeAnalysis.Extensions.IExtensionManager, System.Object, System.Action)
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean)
   at System.Threading.Tasks.Task.Wait(Int32, System.Threading.CancellationToken)
   at Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.SuggestedAction.InvokeWorker(System.Func`1<Microsoft.CodeAnalysis.Document>, Microsoft.CodeAnalysis.Shared.Utilities.IProgressTracker, System.Threading.CancellationToken)
   at Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.SuggestedAction+<>c__DisplayClass18_0.<InvokeCore>b__0()
   at Microsoft.CodeAnalysis.Extensions.IExtensionManagerExtensions.PerformAction(Microsoft.CodeAnalysis.Extensions.IExtensionManager, System.Object, System.Action)
   at Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.SuggestedAction.InvokeCore(System.Func`1<Microsoft.CodeAnalysis.Document>, Microsoft.CodeAnalysis.Shared.Utilities.IProgressTracker, System.Threading.CancellationToken)
   at Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.SuggestedAction.InnerInvoke(Microsoft.CodeAnalysis.Shared.Utilities.IProgressTracker, System.Threading.CancellationToken)
   at Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.FixAllSuggestedAction.InnerInvoke(Microsoft.CodeAnalysis.Shared.Utilities.IProgressTracker, System.Threading.CancellationToken)
   at Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.SuggestedAction+<>c__DisplayClass16_0.<Invoke>b__0(Microsoft.CodeAnalysis.Editor.Host.IWaitContext)
   at Microsoft.VisualStudio.LanguageServices.Implementation.Utilities.VisualStudioWaitIndicator.Wait(System.String, System.String, Boolean, Boolean, System.Action`1<Microsoft.CodeAnalysis.Editor.Host.IWaitContext>)
   at Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.SuggestedAction.Invoke(System.Threading.CancellationToken)
   at Microsoft.VisualStudio.Language.Intellisense.Implementation.LightBulbSession+<>c__DisplayClass39_0.<InvokeAction>b__0(Microsoft.VisualStudio.Language.Intellisense.Utilities.IWaitContext)
   at Microsoft.VisualStudio.Editor.Implementation.VisualStudioWaitIndicator.Wait(System.String, System.String, Boolean, System.Action`1<Microsoft.VisualStudio.Language.Intellisense.Utilities.IWaitContext>)
   at Microsoft.VisualStudio.Language.Intellisense.Implementation.LightBulbSession.InvokeAction(Microsoft.VisualStudio.Language.Intellisense.ISuggestedAction, UInt32)
   at Microsoft.VisualStudio.Language.Intellisense.Implementation.LightBulb+<>c__DisplayClass58_0.<GetMenuItemForAction>b__1()
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(System.Delegate, System.Object, Int32)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(System.Object, System.Delegate, System.Object, Int32, System.Delegate)
   at System.Windows.Threading.DispatcherOperation.InvokeImpl()
   at System.Windows.Threading.DispatcherOperation.InvokeInSecurityContext(System.Object)
   at MS.Internal.CulturePreservingExecutionContext.CallbackWrapper(System.Object)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
   at MS.Internal.CulturePreservingExecutionContext.Run(MS.Internal.CulturePreservingExecutionContext, System.Threading.ContextCallback, System.Object)
   at System.Windows.Threading.DispatcherOperation.Invoke()
   at System.Windows.Threading.Dispatcher.ProcessQueue()
   at System.Windows.Threading.Dispatcher.WndProcHook(IntPtr, Int32, IntPtr, IntPtr, Boolean ByRef)
   at MS.Win32.HwndWrapper.WndProc(IntPtr, Int32, IntPtr, IntPtr, Boolean ByRef)
   at MS.Win32.HwndSubclass.DispatcherCallbackOperation(System.Object)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(System.Delegate, System.Object, Int32)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(System.Object, System.Delegate, System.Object, Int32, System.Delegate)
   at System.Windows.Threading.Dispatcher.LegacyInvokeImpl(System.Windows.Threading.DispatcherPriority, System.TimeSpan, System.Delegate, System.Object, Int32)
   at MS.Win32.HwndSubclass.SubclassWndProc(IntPtr, Int32, IntPtr, IntPtr)

```

**Comment by davkean** [comment]
And sometimes it fixes "all" but the document you are in...

**Comment by davidroth** [comment]
I am seeing the same issue in VS 2015.

**Comment by davkean** [comment]
I attached in the case when it doesn't run, and this stack is throwing OperationCanceledException:

```
 	mscorlib.dll!System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(System.Threading.Tasks.Task task)	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(System.Threading.Tasks.Task task)	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.ConfiguredTaskAwaitable.ConfiguredTaskAwaiter.GetResult()	Unknown
>	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers.ComputeAnalyzerDiagnosticsAsync(Microsoft.CodeAnalysis.Diagnostics.AnalysisScope analysisScope, System.Func<Microsoft.CodeAnalysis.Diagnostics.AsyncQueue<Microsoft.CodeAnalysis.Diagnostics.CompilationEvent>> getEventQueue, int newTaskToken, System.Threading.CancellationToken cancellationToken) Line 624	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers.GetAnalyzerSemanticDiagnosticsCoreAsync(Microsoft.CodeAnalysis.SemanticModel model, Microsoft.CodeAnalysis.Text.TextSpan? filterSpan, System.Collections.Immutable.ImmutableArray<Microsoft.CodeAnalysis.Diagnostics.DiagnosticAnalyzer> analyzers, System.Threading.CancellationToken cancellationToken) Line 586	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers.GetAnalyzerSemanticDiagnosticsAsync(Microsoft.CodeAnalysis.SemanticModel model, Microsoft.CodeAnalysis.Text.TextSpan? filterSpan, System.Collections.Immutable.ImmutableArray<Microsoft.CodeAnalysis.Diagnostics.DiagnosticAnalyzer> analyzers, System.Threading.CancellationToken cancellationToken) Line 564	C#
 	Microsoft.CodeAnalysis.Features.dll!Microsoft.CodeAnalysis.Diagnostics.EngineV2.DiagnosticIncrementalAnalyzer.Executor.ComputeDiagnosticAnalyzerDiagnosticsAsync(Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers analyzerDriverOpt, Microsoft.CodeAnalysis.Document document, Microsoft.CodeAnalysis.Diagnostics.DiagnosticAnalyzer analyzer, Microsoft.CodeAnalysis.Diagnostics.EngineV2.DiagnosticIncrementalAnalyzer.AnalysisKind kind, Microsoft.CodeAnalysis.Text.TextSpan? spanOpt, System.Threading.CancellationToken cancellationToken) Line 407	C#
 	Microsoft.CodeAnalysis.Features.dll!Microsoft.CodeAnalysis.Diagnostics.EngineV2.DiagnosticIncrementalAnalyzer.Executor.ComputeDiagnosticsAsync(Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers analyzerDriverOpt, Microsoft.CodeAnalysis.Document document, Microsoft.CodeAnalysis.Diagnostics.DiagnosticAnalyzer analyzer, Microsoft.CodeAnalysis.Diagnostics.EngineV2.DiagnosticIncrementalAnalyzer.AnalysisKind kind, Microsoft.CodeAnalysis.Text.TextSpan? spanOpt, System.Threading.CancellationToken cancellationToken) Line 140	C#
 	Microsoft.CodeAnalysis.Features.dll!Microsoft.CodeAnalysis.Diagnostics.EngineV2.DiagnosticIncrementalAnalyzer.Executor.GetDocumentAnalysisDataAsync(Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers analyzerDriverOpt, Microsoft.CodeAnalysis.Document document, Microsoft.CodeAnalysis.Diagnostics.EngineV2.DiagnosticIncrementalAnalyzer.StateSet stateSet, Microsoft.CodeAnalysis.Diagnostics.EngineV2.DiagnosticIncrementalAnalyzer.AnalysisKind kind, System.Threading.CancellationToken cancellationToken) Line 78	C#
 	Microsoft.CodeAnalysis.Features.dll!Microsoft.CodeAnalysis.Diagnostics.EngineV2.DiagnosticIncrementalAnalyzer.AnalyzeDocumentForKindAsync(Microsoft.CodeAnalysis.Document document, Microsoft.CodeAnalysis.Diagnostics.EngineV2.DiagnosticIncrementalAnalyzer.AnalysisKind kind, System.Threading.CancellationToken cancellationToken) Line 49	C#
 	[Resuming Async Method]	
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.InvokeMoveNext(object stateMachine)	Unknown
 	mscorlib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx)	Unknown
 	mscorlib.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx)	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.Run()	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.OutputAsyncCausalityEvents.AnonymousMethod__0()	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.ContinuationWrapper.Invoke()	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.TaskAwaiter.OutputWaitEtwEvents.AnonymousMethod__0()	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.ContinuationWrapper.Invoke()	Unknown
 	mscorlib.dll!System.Threading.Tasks.AwaitTaskContinuation.ExecuteWorkItemHelper()	Unknown
 	mscorlib.dll!System.Threading.Tasks.AwaitTaskContinuation.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()	Unknown
 	mscorlib.dll!System.Threading.ThreadPoolWorkQueue.Dispatch()	Unknown
 	mscorlib.dll!System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()	Unknown
 	[Async Call]	
 	Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.Editor.Shared.Preview.PreviewSolutionCrawlerRegistrationServiceFactory.Service.AnalyzeAsync() Line 89	C#
 	[Async Call]	
 	Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.Editor.Shared.Preview.PreviewSolutionCrawlerRegistrationServiceFactory.Service.Unregister(Microsoft.CodeAnalysis.Workspace workspace, bool blockingShutdown) Line 106	C#

**Comment by davkean** [comment]
tag @heejaechang

**Comment by davkean** [comment]
Other times it will show this dialog and just sit there doing nothing (no CPU/disk access):

![image](https://cloud.githubusercontent.com/assets/1103906/21535943/44a2c310-cdd0-11e6-9b8b-49c8ce5a3968.png)

**Comment by heejaechang** [comment]
let me take a look at the part where it asks OOP to return diagnostics and see whether that parts work as expected.

**Comment by Pilchie** [comment]
Moving so @heejaechang tracks it.

**Comment by heejaechang** [comment]
sorry, I dont think I can do this for Rc3

**Comment by heejaechang** [comment]
found reason why sometime nothing happened. in talk with service hub guys to figure out how to handle the issue.

**Comment by CyrusNajmabadi** [comment]
Can you share the information?

**Comment by heejaechang** [comment]
@CyrusNajmabadi it is linked above #16872

**Comment by heejaechang** [comment]
I believe this is now fixed with #16872
